File: app/src/main/java/com/platdmit/simplecloudmanager/app/adapters/ServerViewStateAdapter.java
package com.platdmit.simplecloudmanager.app.adapters;

import android.os.Bundle;

import com.platdmit.simplecloudmanager.app.fragments.ServerTabFragment;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.Lifecycle;
import androidx.viewpager2.adapter.FragmentStateAdapter;

public class ServerViewStateAdapter extends FragmentStateAdapter {

    protected List<ServerTabFragment<? extends Fragment>> mServerTabFragments;
    protected static int mTabCount;
    private long mServerId;

    protected ServerViewStateAdapter(@NonNull FragmentManager fragmentManager, @NonNull Lifecycle lifecycle, long serverId, List<ServerTabFragment<? extends Fragment>> serverTabFragments) {
        super(fragmentManager, lifecycle);
        mServerTabFragments = serverTabFragments;
        mTabCount = serverTabFragments.size();
        mServerId = serverId;
    }

    @NonNull
    @Override
    public Fragment createFragment(int position) {
        Bundle bundle = new Bundle();
        bundle.putLong("ELEMENT_ID", mServerId);
        Fragment fragment = (Fragment) mServerTabFragments.get(position).getInstance();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public int getItemCount() {
        return mTabCount;
    }

    public CharSequence getPageTitle(int position) {
        return mServerTabFragments.get(position).getTitle();
    }

    public static class Builder{
        private int count;
        private final List<ServerTabFragment<? extends Fragment>> mServerTabFragments;

        public Builder(){
            mServerTabFragments = new ArrayList<>();
        }

        public Builder add(ServerTabFragment<? extends Fragment> serverTabFragment){
            mServerTabFragments.add(serverTabFragment);
            count++;
            return this;
        }

        public ServerViewStateAdapter build(FragmentManager fm, Lifecycle lifecycle, long serverId){
            if(count > 0){
                return new ServerViewStateAdapter(fm, lifecycle, serverId, mServerTabFragments);
            } else {
                return null;
            }
        }

    }
}

File: app/src/main/java/com/platdmit/simplecloudmanager/app/adapters/ServerViewStateAdapter.kt
package com.platdmit.simplecloudmanager.app.adapters

import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.Lifecycle
import androidx.viewpager2.adapter.FragmentStateAdapter
import com.platdmit.simplecloudmanager.app.fragments.ServerTabFragment

class ServerViewStateAdapter(
        private val fragmentManager: FragmentManager,
        val lifecycle: Lifecycle,
        val serverId: Long,
        private val serverTabFragments: List<ServerTabFragment<out Fragment>>
) : FragmentStateAdapter(fragmentManager, lifecycle) {

    private val tabCount : Int = serverTabFragments.size

    override fun getItemCount(): Int {
        return tabCount
    }

    override fun createFragment(position: Int): Fragment {
        val fragment : Fragment = serverTabFragments[position].getInstance();
        fragment.arguments = bundleOf("ELEMENT_ID" to serverId)
        return fragment;
    }

    fun getPageTitle(position : Int) : CharSequence = serverTabFragments[position].getTitle()

    companion object {
        private var count : Int = 0;
        private val serverTabFragments : MutableList<ServerTabFragment<out Fragment>> = ArrayList();

        fun add(serverTabFragment: ServerTabFragment<out Fragment>): Companion {
            serverTabFragments.add(serverTabFragment);
            count++
            return this
        }

        fun build(fm : FragmentManager, lifecycle : Lifecycle, serverId: Long) : ServerViewStateAdapter?{
            return if(count > 0){
                ServerViewStateAdapter(fm, lifecycle, serverId, serverTabFragments)
            } else {
                null
            }
        }
    }
}